airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/message/DestinationMessage.kt

```diff
@@ -8,37 +8,72 @@ import com.fasterxml.jackson.databind.JsonNode
 import io.airbyte.cdk.command.DestinationCatalog
 import io.airbyte.cdk.command.DestinationStream
 import io.airbyte.protocol.models.v0.AirbyteMessage
+import io.airbyte.protocol.models.v0.AirbyteRecordMessage
 import io.airbyte.protocol.models.v0.AirbyteStateMessage
 import io.airbyte.protocol.models.v0.AirbyteStreamState
+import io.airbyte.protocol.models.v0.AirbyteStreamStatusTraceMessage
 import io.airbyte.protocol.models.v0.AirbyteStreamStatusTraceMessage.AirbyteStreamStatus
 import io.airbyte.protocol.models.v0.AirbyteTraceMessage
+import io.airbyte.protocol.models.v0.StreamDescriptor
 import jakarta.inject.Singleton
 
 /**
  * Internal representation of destination messages. These are intended to be specialized for
  * usability. Data should be marshalled to these from frontline deserialized objects.
  */
-sealed class DestinationMessage
+sealed interface DestinationMessage {
+    fun asProtocolMessage(): AirbyteMessage
+}
 
 /** Records. */
-sealed class DestinationRecordMessage : DestinationMessage() {
-    abstract val stream: DestinationStream
+sealed interface DestinationStreamAffinedMessage : DestinationMessage {
+    val stream: DestinationStream
 }
 
 data class DestinationRecord(
     override val stream: DestinationStream,
     val data: JsonNode? = null,
     val emittedAtMs: Long,
     val serialized: String
-) : DestinationRecordMessage()
+) : DestinationStreamAffinedMessage {
+    override fun asProtocolMessage(): AirbyteMessage =
+        AirbyteMessage()
+            .withType(AirbyteMessage.Type.RECORD)
+            .withRecord(
+                AirbyteRecordMessage()
+                    .withStream(stream.descriptor.name)
+                    .withNamespace(stream.descriptor.namespace)
+                    .withEmittedAt(emittedAtMs)
+                    .withData(data)
+            )
+}
 
-data class DestinationStreamComplete(
+data class DestinationEndOfStream(
     override val stream: DestinationStream,
-    val emittedAtMs: Long
-) : DestinationRecordMessage()
+    val emittedAtMs: Long,
+    val status: AirbyteStreamStatus,
```

Comment:
I think I prefer the StreamComplete/Incomplete. In general my preference is that anything you're likely to act on differently overall ought to be a case class.

Also the protocol enums maybe ought not leak into these classes.

Reply:
fair points, agree with both of these. I'll switch back to explicit complete/incomplete.

Reply:
I added the complete/incomplete message structs, directly implementing StreamAffinedMessage. Could be convinced to add `sealed interface DestinationEndOfStream: DestinationStreamAffinedMessage`, but it only seemed useful in one place so :shrug: (the `message !is StreamComplete && message !is StreamIncomplete` conditional in ProcessRecordsTask)